### PR TITLE
WT-9199 Checkpoint can miss transactions with commit times before stable

### PR DESCRIPTION
**Pull request by raviprakashgiri29**
This pull request has the following changes.

A csuite Test to reproduce the checkpoint can miss transactions with commit times before stable.
Added a timing stress to the commit transaction.
Added a new commit generation and fixed checkpoint so as not to miss transactions with commit times before stable.

### CODE REVIEW

File: src/txn/txn.c [hunk 1]
```diff
@@ -1978,29 +1988,36 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
     if (!readonly)
         WT_IGNORE_RET(__wt_cache_eviction_check(session, false, false, NULL));
 
-    return (0);
-
+    if (0) {
```
Comment (quchenhao) [inline src/txn/txn.c:1991]: Why this change is needed?

Comment (quchenhao) [reply src/txn/txn.c:1991]: I think it is better to avoid using if (0) if we can.

File: src/txn/txn.c [hunk 2]
```diff
@@ -1711,9 +1712,18 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
     if (!prepare)
         F_CLR(txn, WT_TXN_TS_ROUND_PREPARED);
 
+    /* Enter the commit generation before timestamp validity check. */
+    if (!prepare)
```
Comment (quchenhao) [inline src/txn/txn.c:1716]: I'm OK with this approach.

However, personally I prefer to treat prepare transactions the same as normal transactions. Therefore, the approach to move the code that gets the checkpoint snapshot outside of checkpoint prepare call is a better approach to me.

File: src/txn/txn.c (hunk 1 above)
Comment (raviprakashgiri29) [reply src/txn/txn.c:1991]: This is mainly to avoid calling `__wt_session_gen_leave` twice once in the normal scenario and another in the error scenario.
Another approach to avoid using `if (0)` is to call the rolling back the transaction and other things if `ret != 0`

```
err:
    /*
     * Leave the commit generation after all the updates are processed and the snapshot is released.
     */
    if (!prepare)
        __wt_session_gen_leave(session, WT_GEN_TXN_COMMIT);

    if (ret != 0){
        if (cursor != NULL)
            WT_TRET(cursor->close(cursor));

        if (locked)
            __wt_readunlock(session, &txn_global->visibility_rwlock);

        /* Check for a failure after we can no longer fail. */
        if (cannot_fail)
            WT_RET_PANIC(session, ret,
              "failed to commit a transaction after data corruption point, failing the system");

        /*
         * Check for a prepared transaction, and quit: we can't ignore the error and we can't roll
         * back a prepared transaction.
         */
        if (prepare)
            WT_RET_PANIC(session, ret, "failed to commit prepared transaction, failing the system");

        WT_TRET(__wt_session_reset_cursors(session, false));
        WT_TRET(__wt_txn_rollback(session, cfg));
    }
    return (ret);
```

File: src/txn/txn.c (hunk 2 above)
Comment (raviprakashgiri29) [reply src/txn/txn.c:1716]: @quchenhao I think the scenario we are talking about in this change does not hold good for prepared transactions because there is a check [here](https://github.com/wiredtiger/wiredtiger/blob/develop/src/txn/txn.c#L1967) in the `txn_commit` that the stable timestamp cannot be concurrently increased greater than or equal to the prepared transaction's durable timestamp. So, it is safe to skip the prepared transactions to not enter into the commit generation.

File: src/txn/txn.c (hunk 1 above)
Comment (quchenhao) [reply src/txn/txn.c:1991]: I actually would prefer calling `__wt_session_gen_leave ` twice. But this is mainly a personal preference.

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/include/misc_inline.h [hunk 3]
```diff
@@ -197,6 +197,36 @@ __wt_spin_backoff(uint64_t *yield_count, uint64_t *sleep_usecs)
     __wt_sleep(0, (*sleep_usecs));
 }
 
+/*
+ * __wt_timing_stress_sleep_for_secs --
+ *     Optionally add delay to stress code paths. Wait for the specified amount of time.
+ */
+static inline void
```
[the remark this answers is not in the document]
Comment (bvpvamsikrishna) [reply src/include/misc_inline.h:204]: Better to combine this and the next function into one.

[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/include/misc_inline.h:204]: Every module has a specific function like this. For example, `checkpoint_timing_stress` accepts time as an argument.

As Vamsi suggested, combining these two functions will be good. If the user doesn't provide the time, use the random one as it exists.

[the remark this answers is not in the document]
Comment (raviprakashgiri29) [reply src/include/misc_inline.h]: @bvpvamsikrishna @kommiharibabu Fixed in c41b9ebfd857133657e4b3725d71371d135b31a1

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

**Comment by raviprakashgiri29** [comment]
Here is the new pull request: https://github.com/wiredtiger/wiredtiger/pull/10323
